### DISCUSSION TITLE
MVVM中的Ribbon控件问题

### DISCUSSION THREAD

**Discussion by tsctsc6** [BYJRK/DotNet-Discussions#269] [MVVM]
具体问题描述和源代码，请看我的[仓库](https://github.com/tsctsc6/RibbonMVVM)

**Comment by CodingOctocat** [comment]
- 第一个问题，`TextBox` 的 `Text` 绑定默认是失去焦点才更新，这导致你如果按 `Ribbon` 的按钮，`Text` 尚未更新绑定源：
```xaml
Text="{Binding InputText}"
```
修复：显式将 `UpdateSourceTrigger` 设置为 `PropertyChanged`，实时更新绑定源
```xaml
Text="{Binding InputText, UpdateSourceTrigger=PropertyChanged}"
```

---
- 第二个问题，根据绑定错误可知，找不到 `RibbonWindow`：
```
System.Windows.Data Error: 4 : Cannot find source for binding with reference 'RelativeSource FindAncestor, AncestorType='System.Windows.Controls.Ribbon.RibbonWindow', AncestorLevel='1''. BindingExpression:Path=WindowState; DataItem=null; target element is 'Ribbon' (Name=''); target property is 'NoTarget' (type 'Object')
System.Windows.Data Error: 4 : Cannot find source for binding with reference 'RelativeSource FindAncestor, AncestorType='System.Windows.Controls.Ribbon.RibbonWindow', AncestorLevel='1''. BindingExpression:Path=IsActive; DataItem=null; target element is 'Ribbon' (Name=''); target property is 'NoTarget' (type 'Object')
```
修复：将 `MainWindow` 类型更改为 `RibbonWindow`
```xaml
<RibbonWindow x:Class="Ribbon_MVVM.MainWindow">
   ...
</RibbonWindow>
```
```csharp
public partial class MainWindow : RibbonWindow { ... }
```

**Reply by tsctsc6** [reply]
谢谢

**Comment by tsctsc6** [comment]
但是使用了RibbonWindow后，标题的文字分辨率变低了，有解决办法吗
![image](https://github.com/BYJRK/DotNet-Discussions/assets/93368014/06e26398-a076-4db3-8478-76a9b3963707)

**Reply by CodingOctocat** [reply]
RibbonWindow 设置：
```xaml
UseLayoutRounding="True"
SnapsToDevicePixels="True"
```
不过它的标题好像没有垂直居中。

**Reply by tsctsc6** [reply]
谢谢。但确实，使用了RibbonWindow后，最大化窗口的话，窗口标题有一部分会出屏幕；并且TextBox的Margin为“5 5 5 5”，TextBox周围的空白也没有5那么多了。
![image](https://github.com/BYJRK/DotNet-Discussions/assets/93368014/792e99a0-0364-4743-b72c-43b3b920967a)

**Comment by tsctsc6** [comment]
如果使用Ribbon按钮，并且对性能有点需求或者对TextBox的实时性没有要求，我们能否把思路逆转过来，想办法在点击Ribbon按钮后，使TextBox失去焦点，从而更新绑定源？

**Reply by CodingOctocat** [reply]
在命令方法中调用 `OnPropertyChanged(nameof(InputText))` 试试。

**Reply by tsctsc6** [reply]
不行，甚至会把TextBox中的文字回到上一次的状态；看来这是由InputText到前台控件方向

**Reply by tsctsc6** [reply]
或者激进一点，```UpdateSourceTrigger=Explicit```，在ViewModel中手动控制

**Reply by CodingOctocat** [reply]
有一种解决方式是设置 `Ribbon` 的 `FocusManager.IsFocusScope="False"`，然后添加一个 `PreviewMouseDown` 事件，执行 `ribbon.Focus();` 代码，但将 `IsFocusScope` 设置为 `False` 有什么副作用我就没了解了。

**Reply by tsctsc6** [reply]
好的，谢谢
